File: #48 - ARBOL DE NAVIDAD/JesusWay69.py
import os, platform, random

if (platform.platform().startswith("macOS") or platform.platform().startswith("Linux")):
    os.system('clear')
else:
    os.system('cls')


""" * EJERCICIO:
 * ¡Ha comenzado diciembre! Es hora de montar nuestro
 * árbol de Navidad...
 *
 * Desarrolla un programa que cree un árbol de Navidad
 * con una altura dinámica definida por el usuario por terminal.
 *
 * Ejemplo de árbol de altura 5 (el tronco siempre será igual):
 *
 *     *
 *    ***
 *   *****
 *  *******
 * *********
 *    |||
 *    |||
 *
 * El usuario podrá seleccionar las siguientes acciones:
 * 
 * - Añadir o eliminar la estrella en la copa del árbol (@)
 * - Añadir o eliminar bolas de dos en dos (o) aleatoriamente
 * - Añadir o eliminar luces de tres en tres (+) aleatoriamente
 * - Apagar (*) o encender (+) las luces (conservando su posición)
 * - Una luz y una bola no pueden estar en el mismo sitio
 *
 * Sólo puedes añadir una estrella, y tantas luces o bolas
 * como tengan cabida en el árbol. El programa debe notificar
 * cada una de las acciones (o por el contrario, cuando no
 * se pueda realizar alguna)."""

def create_tree(hight:int)->list:
    tree = []
    base = hight * 2 - 1
    branch = 1
    for i in range(hight):
        tree.append([])
        [tree[i].append(' ') for j in range(base // 2)]
        [tree[i].append('*') for k in range(branch)]
        [tree[i].append(' ') for l in range(base // 2)]
        base -= 2
        branch += 2
    for m in range (1, 3):
        tree.append([])
        [tree[i + m].append(' ') for n in range(hight - 2)]
        [tree[i + m].append('|') for o in range(3)]
        [tree[i + m].append(' ') for p in range(hight - 2)]

    return tree

def show_tree(tree:list):
    print()
    for row in tree:
        for column in row:
            print(column, end='')
        print()

def add_star(tree:list, switch:bool)->list:
    if switch:
        star = ['@' if x == '*' else x for x in tree[0]]
    else:
        star = ['*' if x == '@' else x for x in tree[0]]
    tree[0] = star
    return tree

def add_balls(tree:list):
    i=0
    while i !=2:
        branch = random.randint(1, len(tree)-3)
        ball = random.randint(0, len(tree[branch]))
        print("ronda:",i," rama:",branch," columna:", ball)
        if tree[branch][ball-1] != '*':
            continue
        else:
            tree[branch][ball-1] = 'o'
            i+=1
    return tree

def remove_balls(tree:list):
    for row in range(1, len(tree)-2):
        for column in range(len(tree[row])):
            if tree[row][column] == 'o':
                tree[row][column] = '*'
    return tree




# tree = create_tree(15)
# show_tree(tree)
# show_tree(add_star(tree, True))
# show_tree(add_balls(tree))
# show_tree(add_balls(tree))
# show_tree(remove_balls(tree))



while True:
    print("""
    
    1- Crear árbol
    2- Añadir 2 bolas aleatoriamente
    3- Quitar todas las bolas
    4- Añadir 3 luces aleatoriamente 
    5- Encender las luces
    6- Apagar las luces
          
    """)

    option = input("Selecciona una opción del 1 al 6: ") 

    match option:
        case "1":
            hight = input("Indroduzca la altura del árbol a crear: ")
            tree = create_tree(hight)
        case "2":
            add_balls(tree=None)
        case "3":
            remove_balls(tree=None)
        case "4":
            pass
        case "5":
            pass
        case "6":
            pass

File: #48 - ARBOL DE NAVIDAD/JesusWay69.java
package reto_49;

import java.util.Scanner;

public class JesusWay69 {

    public static void main(String[] args) {
        //char [][] myTree = createTree(5);
        /*showTree(myTree);
        topStar(myTree, true);
        showTree(myTree);
        topStar(myTree, false);
        showTree(myTree);*/
        char[][] myTree = null;

        do {
            Scanner sc = new Scanner(System.in);
            System.out.println("""
              1- Crear árbol
              2- Añadir estrella 
              3- Eliminar estrella 
              4- Añadir 2 bolas aleatoriamente
              5- Quitar todas las bolas
              6- Añadir 3 luces aleatoriamente 
              7- Encender las luces
              8- Apagar las luces
                             """);
            System.out.print("Elija una opción: ");
            Byte option = sc.nextByte();
            if (option>8 || option <1 || option instanceof Byte == false){
                System.out.println("El dato tiene que ser numérico del 1 al 8");
            }
            switch (option) {
                case 1:
                    System.out.println("Introduzca la altura del árbol (mayor de 3): ");
                    int hight = sc.nextInt();
                    myTree = createTree(hight);
                    showTree(myTree);
                    break;
                case 2:
                    topStar(myTree, true);
                    showTree(myTree);
                    break;
                case 3:
                    topStar(myTree, false);
                    showTree(myTree);
                    break;
                case 4:
                    addBalls(myTree);
                    showTree(myTree);
                    break;
                case 5:
                case 6:
                case 7:
                case 8:
                default:

            }
        } while (true);
    }

    public static char[][] createTree(int files) {
        int columns = files * 2 - 1;
        int branch = 1;
        char[][] tree = new char[files + 2][columns];

        if (files <= 3) {
            System.out.println("No se puede crear un árbol de menos de 4 alturas");
            tree = null;
            return tree;
        }

        for (int i = 0; i < files; i++) {
            for (int j = 0; j < columns / 2; j++) {
                tree[i][j] = ' ';

            }
            for (int k = columns / 2; k < columns / 2 + branch; k++) {
                tree[i][k] = '*';

            }
            for (int l = columns / 2 + branch; l < files * 2 - 1; l++) {
                tree[i][l] = ' ';

            }
            columns -= 2;
            branch += 2;
        }
        columns = files * 2 - 1;
        for (int m = 0; m < 2; m++) {
            for (int n = 0; n < files - 2; n++) {
                tree[files + m][n] = ' ';

            }
            for (int p = files - 2; p < files + 1; p++) {
                tree[files + m][p] = '|';

            }
            for (int q = files + 1; q < files * 2 - 1; q++) {
                tree[files + m][q] = ' ';

            }

        }
        //System.out.println(Arrays.deepToString(tree));
        return tree;
    }

    public static void showTree(char[][] tree) {
        int i = 0;
        for (char[] branch : tree) {
            i++;
            for (char c : branch) {
                System.out.print(c);
            }
            System.out.println();

        }
    }

    public static char[][] topStar(char[][] tree, boolean switchStar) {
        if (tree == null) {
            System.out.println("ERROR: Hay que crear un árbol antes de poder modificarlo (opción 1)");
            return null;
        } else if (switchStar) {
            tree[0][tree[0].length / 2] = '@';
        } else {
            tree[0][tree[0].length / 2] = '*';
        }

        return tree;
    }

    public static char[][] addBalls(char[][] tree) {

        for (int i = 0; i < 2; i++) {
            int randomFile = (int) (Math.random() * tree.length - 2 + 1);
            int randomColumn = (int) (Math.random() * tree[0].length);
            if (tree[randomFile][randomColumn] == '*') {
                tree[randomFile][randomColumn] = 'O';
            } else {
                i--;
            }
        }
        return tree;
    }

}
